lib/SPIRV/SPIRVLowerMemmove.cpp

```diff
@@ -54,54 +54,51 @@ namespace SPIRV {
 
 class SPIRVLowerMemmoveBase {
 public:
-  SPIRVLowerMemmoveBase() : Context(nullptr), Mod(nullptr) {}
+  SPIRVLowerMemmoveBase() : Context(nullptr) {}
+
   void LowerMemMoveInst(MemMoveInst &I) {
+    // There is no direct equivalent of @llvm.memmove in SPIR-V and the closest
+    // instructions are 'OpMemoryCopy' and 'OpMemoryCopySized'.
+    //
+    // 'OpMemoryCopy' does not accept amount of bytes to copy and infers that
+    // from type which is being copied; also it only allows to copy value of a
+    // particular type to pointer pointing to the same type.
+    //
+    // 'OpMemoryCopySized' is closer to @llvm.memmove, because it actually
+    // copies bytes, but unlike memove it doesn't support overlapping source
+    // and destination. Therefore, we replace memmove with two
+    // 'OpMemoryCopySized' instructions: the first one copies bytes from source
+    // to a temporary location, the second one copies bytes from that temporary
+    // location to the destination.
     IRBuilder<> Builder(I.getParent());
     Builder.SetInsertPoint(&I);
-    auto *Dest = I.getRawDest();
-    auto *Src = I.getRawSource();
-    if (isa<PHINode>(Src))
-      report_fatal_error("llvm.memmove of PHI instruction result not supported",
-                         false);
-    // The source could be bit-cast or addrspacecast from another type,
-    // need the original type for the allocation of the temporary variable
-    auto *SrcTy = Src->stripPointerCasts()->getType();
+
     auto *Length = cast<ConstantInt>(I.getLength());
-    MaybeAlign Align = I.getSourceAlign();
-    auto Volatile = I.isVolatile();
-    Value *NumElements = nullptr;
-    uint64_t ElementsCount = 1;
-    if (SrcTy->isArrayTy()) {
-      ElementsCount = SrcTy->getArrayNumElements();
-      NumElements = Builder.getInt32(ElementsCount);
-    }
-    // Get number of bits to move and allocate memory appropriately:
-    // if lenght is bigger than a pointer base type size, then create an
-    // alloca of an array type with the same base type.
-    const uint64_t LenBits = Length->getZExtValue();
-    const uint64_t LayoutTypeBites =
-        Mod->getDataLayout().getTypeSizeInBits(SrcTy->getPointerElementType()) *
-        ElementsCount;
-    auto *AllocaTy = SrcTy->getPointerElementType();
-    if (LenBits > LayoutTypeBites) {
-      const uint64_t ArraySize = LenBits / LayoutTypeBites;
-      AllocaTy = ArrayType::get(SrcTy->getPointerElementType(), ArraySize);
-    }
-    auto *Alloca = Builder.CreateAlloca(AllocaTy, NumElements);
-    if (Align.hasValue()) {
-      Alloca->setAlignment(Align.getValue());
-    }
+    auto *AllocaTy = ArrayType::get(IntegerType::getInt8Ty(*Context),
+                                    Length->getZExtValue());
+    MaybeAlign SrcAlign = I.getSourceAlign();
+
+    // FIXME: What if @llvm.memmove located within a loop? Wouldn't this be a
+    // huge memory eater?
```

Comment:
Perhaps lifetime intrinsics may give a hint to optimizer, so it could re-use the memory.

Reply:
Thanks for pointing that out, indeed lifetime intrinsics should deal with this. However, as a first argument they accept the allocation size and we do not pass anything there, like the allocation is variable-sized, which I guess is not the case. But the consequences of that are not clear to me

Reply:
Replaced this FIXME with another one in dbc54dd752c04bb65e8d2e2aedaaac22c247266e